File: automaticEditing.cpp

/*
 * automatic editing
 * http://uva.onlinejudge.org/index.php?option=com_onlinejudge&Itemid=8&page=show_problem&problem=1056
 *
 * Perceptive Software Code Rush 2012 C++
 * 2/7/13 status: complete
 */
#include <iostream>
#include <fstream>
#include <utility>	//pair
#include <vector>	
#include <string>	//string replace and find
using namespace std;

int main()	{
	
	fstream fin;
	fin.open("autoedit.in", fstream::in);
	
	int rules;
	string buffer, key, value, meat;
	vector<pair <string, string> > m;	
	vector<pair <string, string> >::iterator it;
	getline(fin, buffer);
	rules = atoi(buffer.c_str());
	
	while(rules != 0)	{

		//Grabbing rules
		while(rules != 0)	{
			getline(fin, key);
			getline(fin, value);
			m.push_back(pair<string, string>(key, value));
			rules--;	
		}
		//Chopping up the sentence
		getline(fin, meat);
		
		for(it = m.begin(); it != m.end(); it++)	{
			while( meat.find(it->first) != string::npos)	{
				meat = meat.replace(meat.find(it->first), it->first.length(), it->second);		
			}
		}
		cout << meat << endl;

		//clearing vector
		m.clear();
		getline(fin, buffer);
		rules = atoi(buffer.c_str());
	}	
	fin.close();

}

File: automaticEditing.java
/*
 * Automatic Editing
 * Perceptive Software Code Rush 2012 Problem 9
 * http://uva.onlinejudge.org/index.php?option=com_onlinejudge&Itemid=8&page=show_problem&problem=1056
 *
 * Status: COMPLETE 
 * Victoria Wu, Marvin Bellamy
 * 2/6/13
 */

import java.io.*;
import java.util.*;	//scanner, hashmap

//pair class modified from
//http://stackoverflow.com/questions/156275/what-is-the-equivalent-of-the-c-pairl-r-in-java

class Pair<A, B>	{
	public A first; public B second;
	public Pair(A one, B two)	{first = one; second =two;}
}

public class automaticEditing	{
	public static void main(String args[]) throws IOException	{
		
		Scanner in = new Scanner(new FileReader("autoedit.in"));
		
		//Be careful when intermixing next and nextline. Next will only take up to \n	
		int rules = Integer.parseInt(in.nextLine());
		String key, value;
		String meat;
		ArrayList<Pair<String, String> > m = new ArrayList<Pair<String, String>>();
		while(rules != 0)	{

			//Getting the rules
			while(rules != 0)	{
				key = in.nextLine();
				value = in.nextLine();
				rules--;
				m.add(new Pair<String, String>(key, value));
				//System.out.println("READ IN: " + key + " = " + value);
			}	

			//Chopping up the sentence
			meat = in.nextLine();	
			//System.out.println("MEAT: " + meat);
			for(Pair<String, String> entry : m)	{	
				//System.out.println("Applying rule: " + entry.first + " = " + entry.second);
				while(meat.indexOf(entry.first) != -1)	{
					meat = meat.replaceFirst(entry.first, entry.second);
				}		
			}
			System.out.println(meat);

			//clearing array list
			m.clear();
			rules = Integer.parseInt(in.nextLine());

		}
		in.close();
	}
}
